src/renderer/atlas/BackendD3D.cpp

```diff
@@ -1299,15 +1355,31 @@ bool BackendD3D::_drawGlyph(const RenderingPayload& p, const BackendD3D::AtlasFo
         static_cast<f32>(rect.y - bt),
     };
 
-    if (transform)
+    _d2dBeginDrawing();
+
+    if (isBoxGlyph)
     {
-        auto& t = *transform;
-        t.dx = (1.0f - t.m11) * baselineOrigin.x;
-        t.dy = (1.0f - t.m22) * baselineOrigin.y;
-        _d2dRenderTarget->SetTransform(&t);
+        const D2D1_RECT_F clipRect{
+            static_cast<f32>(rect.x) / transform.m11,
+            static_cast<f32>(rect.y) / transform.m22,
+            static_cast<f32>(rect.x + rect.w) / transform.m11,
+            static_cast<f32>(rect.y + rect.h) / transform.m22,
+        };
+        _d2dRenderTarget4->PushAxisAlignedClip(&clipRect, D2D1_ANTIALIAS_MODE_ALIASED);
     }
+    const auto boxGlyphCleanup = wil::scope_exit([&]() {
+        if (isBoxGlyph)
+        {
+            _d2dRenderTarget4->PopAxisAlignedClip();
+        }
+    });
 
-    _d2dBeginDrawing();
+    if (needsTransform)
+    {
+        transform.dx = (1.0f - transform.m11) * baselineOrigin.x;
+        transform.dy = (1.0f - transform.m22) * baselineOrigin.y;
+        _d2dRenderTarget->SetTransform(&transform);
+    }

```

Comment:
This code moved down from line 1302, because calculating the `PushAxisAlignedClip` rectangle is much simpler that way (the transform is still in absolute coordinates, instead of being relative to the baseline origin of the glyph).